reactor-core/src/main/java/reactor/core/publisher/FluxBuffer.java

```diff
@@ -477,6 +481,17 @@ public void onNext(T t) {
 				return;
 			}
 
+			boolean added = isEmpty();
+			for (C b : this) {
+				added |= b.add(t);
+			}
+
+			if (!added) {
+				Operators.onDiscard(t, actual.currentContext());
+				s.request(1);
```

Comment:
I do have doubts here - the demand flow without Set is that every buffer that is initiated has a chance to be fulfilled. If the downstream requests 1 it means a full buffer should be completed. As the buffers are delivered, the downstream can request more. With potentially non-deterministic implementations of Collection which can unpredictably return false some assumptions might not hold. There are a few assumptions here that dictate the flow that follows. I'd probably prefer to back away from implementations for `skip != maxSize` as they become complicated to reason about and explain to users what are the risks (e.g. keeping a growing list of undelivered buffers). I think we could terminate early instead of stalling and deliver an error to the downstream and cancel the source in case the buffer is unable to accept the item.

Reply:
Regarding the early termination, I clarified in the general comment - better to leave the undefined behaviour in such case.

Reply:
> the demand flow without Set is that every buffer that is initiated has a chance to be fulfilled. If the downstream requests 1 it means a full buffer should be completed. As the buffers are delivered, the downstream can request more. With potentially non-deterministic implementations of Collection which can unpredictably return false some assumptions might not hold.

Indeed, this is one of the reasons I originally implemented this with "if the data can't be added to all in-flight buffers, then discard it", which might avoid some of this weirdness.

> I'd probably prefer to back away from implementations for skip != maxSize as they become complicated to reason about and explain to users what are the risks ... better to leave the undefined behaviour in such case.

I agree 😄